### PR TITLE
Allow pagination parameters to be passed through select option

### PR DESCRIPTION
**Pull request by garymardell**
A problem faced is with pagination of `has many` relationships. The root nodes are paginated using a cursor based paginator where as the nested `has many` relationships are paginated with an offset paginator. This means passing the cursor from a has many connection back to a root query will raise an error. 

## Example

We want to paginate through all the comments for a given post. 

```graphql
query PostWithComments($id: GadgetID!) {
  post(id: $id) {
    id
    comments {
      pageInfo {
        endCursor
        hasNextPage
      }
      edges {
        node {
          id
        }
      }
    }
  }
}
```

Passing the `endCursor` to a `findMany` will fail. Instead we need to get the next page of comments for this given post. 

```graphql
query PostWithComments($id: GadgetID!) {
  post(id: $id) {
    id
    comments(after: "cursor") {
      pageInfo {
        endCursor
        hasNextPage
      }
      edges {
        node {
          id
        }
      }
    }
  }
}
```

## Solution

Technically this is already possible to do today with the API however you need to use the `tiny-query-graph-compiler` directly with the `Call`. This change introduces some specific keys that if present are automatically extracted into the `Call`.  

## PR Checklist

- [ ] Important or complicated code is tested
- [ ] Any user facing changes are documented in the Gadget-side changelog
- [ ] Any immediate changes are slated for release in Gadget via a generated package dependency bump
- [ ] Versions within this monorepo are matching and there's a valid upgrade path

### CODE REVIEW

File: packages/api-client-core/src/FieldSelection.ts [hunk 1]
```diff
@@ -3,5 +3,28 @@
  * Example: `{ id: true, name: false, richText: { markdown: true, html: false } }`
  **/
 export interface FieldSelection {
-  [key: string]: boolean | null | undefined | FieldSelection;
+  [key: string]: FieldSelectionType;
```
Comment (garymardell) [inline packages/api-client-core/src/FieldSelection.ts:6]: In order to support `after` and `first` parameters we need to allow the user to pass through both strings and numbers values. Adding them directly to this type means that all fields would accept them which is not ideal. In order to mitigate this I have introduced a specific `ConnectionFieldSelection` that has the just pagination parameters and edges.

File: packages/api-client-core/src/FieldSelection.ts [hunk 2]
```diff
@@ -3,5 +3,28 @@
...
+  pageInfo?: FieldSelection;
+  edges: EdgeFieldSelection;
+}
+
+export interface EdgeFieldSelection {
+  node: FieldSelection;
+}
+
+export function isFieldSelection(selection: FieldSelectionType): selection is FieldSelection {
+  return !!(selection && typeof selection == "object" && !("edges" in selection));
```
Comment (angelini) [inline packages/api-client-core/src/FieldSelection.ts:25]: Will this cause us to break `compileSelect` if they add a model field called `edges`?

Comment (angelini) [reply packages/api-client-core/src/FieldSelection.ts:25]: Which we already have 2 models in prod which would have hit this case.

Comment (garymardell) [reply packages/api-client-core/src/FieldSelection.ts]: Good catch, this broke a specific case when a `BelongsTo` has a node called edges. I have added additional check that also tests that if there is a field called `edges` in the selection that it is also an object (indicating that it is selecting a connection). I have also added a bunch of additional tests for each of the cases.

**Review by angelini — APPROVED** [review]
Awesome :+1:

**Comment by airhorns** [comment]
I think this approach needs a bit more thinking, it means that you can't have a field named `after` when you have a pagination argument named `after` also. I think we need a different spot for the args to go.

**Comment by garymardell** [comment]
Hmm I think almost all routes here with it being part of `select` lead to issues with keys being excluded or the types being incredibly awkward.

**Pull request closed without merging**